### ISSUE TITLE
Test count_is_big_endian sometimes fails

### ISSUE THREAD

**Issue by dtolnay** [zonyitoo/bson-rs#38]
It seems to work 2 out of 3 times. It fails with:

```
thread 'oid::count_is_big_endian'
panicked at 'assertion failed: `(left == right)` (left: `17`, right: `0`)', src/oid.rs:333
```

**Comment by dtolnay** [comment]
It works consistently with `RUST_TEST_THREADS=1 cargo test` so there is a race condition somewhere when tests are run in parallel.

**Comment by kyeah** [comment]
If I had to guess, `count_generation` and `count_is_big_endian` are conflicting by setting `OID_COUNTER` to different start values. 

Looking at them now, I'm not sure if they are even testing different things -- we could probably combine them into one test and avoid having to restructure the tests/struct.

**Comment by dtolnay** [comment]
Some more failures:

```
---- oid::count_is_big_endian stdout ----
	thread 'oid::count_is_big_endian' panicked at 'assertion failed: `(left == right)` (left: `51`, right: `52`)', src/oid.rs:335
note: Run with `RUST_BACKTRACE=1` for a backtrace.

---- oid::count_generation stdout ----
	thread 'oid::count_generation' panicked at 'assertion failed: `(left == right)` (left: `52222`, right: `1122867`)', src/oid.rs:322
```

**Comment by kyeah** [comment]
Yup, would definitely seem like they're conflicting. I'm pretty busy in the next couple of weeks, but if you or anyone else wants to take on combining both tests or finding another option to avoid setting a shared OID_COUNTER, let me know and go for it.

**Comment by saghm** [comment]
Given that `OID_COUNTER` will never actually be initialized with two different starting values in practice, would it be an issue if the tests were changed to use the same initial count? Barring that, would it acceptable to just enforce that our tests should be run single-threaded? (We have to do that in the Mongo driver due to similar issues with accessing the database in parallel).

**Comment by kyeah** [comment]
The issue would remain, since they both increment OID_COUNTER, you'll never be sure if the counter is at the initial count or the other test came in and incremented after you set it.

 Considering this is the only case where we have a parallel issue, I think we should strive for parallel tests; the mongo tests are much more complex since they deal with live databases.

**Comment by saghm** [comment]
That makes sense. One thing I noticed when perusing the code was [the comment about static mutexes](https://github.com/zonyitoo/bson-rs/blob/master/src/oid.rs#L254). I checked [the docs](https://doc.rust-lang.org/std/sync/struct.StaticMutex.html) to see if it had stabilized since then, and was surprised to notice that it actually got deprecated in favor of using the [`lazy_static` crate](https://github.com/rust-lang-nursery/lazy-static.rs), which lets you make static variables using some fancy typing tricks. I think using `lazy_static` to wrap the counter in an `RwLock` (not a `Mutex`, or else acquiring the lock on it stops the rest of the test from being able to proceed) might fix the problem, but I haven't had a chance to actually implement it yet and test it out.
